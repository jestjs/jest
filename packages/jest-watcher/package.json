{
  "name": "jest-watcher",
  "description": "Delightful JavaScript Testing.",
  "version": "30.0.0-alpha.7",
  "main": "./build/index.js",
  "types": "./build/index.d.ts",
  "exports": {
    ".": {
      "types": "./build/index.d.ts",
      "require": "./build/index.js",
      "import": "./build/index.mjs",
      "default": "./build/index.js"
    },
    "./package.json": "./package.json"
  },
  "dependencies": {
    "@jest/test-result": "workspace:*",
    "@jest/types": "workspace:*",
    "@types/node": "*",
    "ansi-escapes": "^4.2.1",
    "chalk": "^4.0.0",
    "emittery": "^0.13.1",
    "jest-util": "workspace:*",
    "string-length": "^4.0.1"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/jestjs/jest.git",
    "directory": "packages/jest-watcher"
  },
  "bugs": {
    "url": "https://github.com/jestjs/jest/issues"
  },
  "engines": {
    "node": "^16.10.0 || ^18.12.0 || >=20.0.0"
  },
  "homepage": "https://jestjs.io/",
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  }
}
