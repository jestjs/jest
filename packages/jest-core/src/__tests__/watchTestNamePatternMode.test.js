/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import chalk from 'chalk';
import {TestPathPatterns} from '@jest/pattern';
import {KEYS} from 'jest-watcher';

const runJestMock = jest.fn();

jest.mock(
  '../SearchSource',
  () =>
    class {
      findMatchingTests(pattern) {
        return {paths: []};
      }
    },
);

jest.doMock('chalk', () => new chalk.Instance({level: 0}));

jest.doMock('strip-ansi');
require('strip-ansi').mockImplementation(str => str);

jest.doMock(
  '../runJest',
  () =>
    function () {
      const args = [...arguments];
      const [{onComplete}] = args;
      runJestMock.apply(null, args);

      // Call the callback
      onComplete({
        snapshot: {},
        testResults: [
          {
            testResults: [{title: 'should return the correct index when'}],
          },
          {
            testResults: [{title: 'should allow test siblings to modify'}],
          },
          {
            testResults: [{title: 'might get confusing'}],
          },
          {
            testResults: [
              {title: 'should handle length properties that cannot'},
            ],
          },
          {
            testResults: [{title: 'should recognize various types'}],
          },
          {
            testResults: [{title: 'should recognize null and undefined'}],
          },
          {
            testResults: [{title: 'should not output colors to pipe'}],
          },
          {
            testResults: [{title: 'should convert string to a RegExp'}],
          },
          {
            testResults: [
              {title: 'should escape and convert string to a RegExp'},
            ],
          },
          {
            testResults: [{title: 'should convert grep string to a RegExp'}],
          },
        ],
      });

      return Promise.resolve();
    },
);

const watch = require('../watch').default;

const globalConfig = {
  rootDir: '',
  testPathPatterns: new TestPathPatterns([]),
  watch: true,
};

afterEach(runJestMock.mockReset);

describe('Watch mode flows', () => {
  let pipe;
  let hasteMapInstances;
  let contexts;
  let stdin;

  beforeEach(() => {
    pipe = {write: jest.fn()};
    hasteMapInstances = [{on: () => {}}];
    contexts = [{config: {}}];
    stdin = new MockStdin();
  });

  it('Pressing "T" enters pattern mode', () => {
    contexts[0].config = {rootDir: ''};
    watch(globalConfig, contexts, pipe, hasteMapInstances, stdin);

    // Write a enter pattern mode
    stdin.emit('t');
    expect(pipe.write).toHaveBeenCalledWith(' pattern â€º ');

    const assertPattern = hex => {
      pipe.write.mockReset();
      stdin.emit(hex);
      expect(pipe.write.mock.calls.join('\n')).toMatchSnapshot();
    };

    // Write a pattern
    for (const pattern of ['c', 'o', 'n', ' ', '1', '2'])
      assertPattern(pattern);

    for (const pattern of [KEYS.BACKSPACE, KEYS.BACKSPACE])
      assertPattern(pattern);

    for (const pattern of ['*']) assertPattern(pattern);

    // Runs Jest again
    runJestMock.mockReset();
    stdin.emit(KEYS.ENTER);
    expect(runJestMock).toHaveBeenCalled();

    // globalConfig is updated with the current pattern
    expect(runJestMock.mock.calls[0][0].globalConfig).toMatchObject({
      onlyChanged: false,
      testNamePattern: 'con *',
      watch: true,
      watchAll: false,
    });
  });
});

class MockStdin {
  constructor() {
    this._callbacks = [];
  }

  setRawMode() {}

  resume() {}

  setEncoding() {}

  on(evt, callback) {
    this._callbacks.push(callback);
  }

  emit(key) {
    for (const cb of this._callbacks) cb(key);
  }
}
