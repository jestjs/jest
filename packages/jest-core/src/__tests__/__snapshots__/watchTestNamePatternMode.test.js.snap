// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Watch mode flows Pressing "T" enters pattern mode 1`] = `
"<eraseLine>
<moveCursorToColumn1>
<eraseScreenDown>
 pattern › c
<saveCursorPosition>
<moveCursorToRow6Column13>
<restoreCursorPosition>
"
`;

exports[`Watch mode flows Pressing "T" enters pattern mode 2`] = `
"<eraseLine>
<moveCursorToColumn1>
<eraseScreenDown>
 pattern › co
<saveCursorPosition>
<moveCursorToRow6Column14>
<restoreCursorPosition>
"
`;

exports[`Watch mode flows Pressing "T" enters pattern mode 3`] = `
"<eraseLine>
<moveCursorToColumn1>
<eraseScreenDown>
 pattern › con
<saveCursorPosition>
<moveCursorToRow6Column15>
<restoreCursorPosition>
"
`;

exports[`Watch mode flows Pressing "T" enters pattern mode 4`] = `
"<eraseLine>
<moveCursorToColumn1>
<eraseScreenDown>
 pattern › con 
<saveCursorPosition>
<moveCursorToRow6Column16>
<restoreCursorPosition>
"
`;

exports[`Watch mode flows Pressing "T" enters pattern mode 5`] = `
"<eraseLine>
<moveCursorToColumn1>
<eraseScreenDown>
 pattern › con 1
<saveCursorPosition>
<moveCursorToRow6Column17>
<restoreCursorPosition>
"
`;

exports[`Watch mode flows Pressing "T" enters pattern mode 6`] = `
"<eraseLine>
<moveCursorToColumn1>
<eraseScreenDown>
 pattern › con 12
<saveCursorPosition>
<moveCursorToRow6Column18>
<restoreCursorPosition>
"
`;

exports[`Watch mode flows Pressing "T" enters pattern mode 7`] = `
"<eraseLine>
<moveCursorToColumn1>
<eraseScreenDown>
 pattern › con 1
<saveCursorPosition>
<moveCursorToRow6Column17>
<restoreCursorPosition>
"
`;

exports[`Watch mode flows Pressing "T" enters pattern mode 8`] = `
"<eraseLine>
<moveCursorToColumn1>
<eraseScreenDown>
 pattern › con 
<saveCursorPosition>
<moveCursorToRow6Column16>
<restoreCursorPosition>
"
`;

exports[`Watch mode flows Pressing "T" enters pattern mode 9`] = `
"<eraseLine>
<moveCursorToColumn1>
<eraseScreenDown>
 pattern › con *
<saveCursorPosition>
<moveCursorToRow6Column17>
<restoreCursorPosition>
"
`;
